runtime/src/iree/hal/drivers/hip/hip_device.c

```diff
@@ -1031,64 +1379,81 @@ static iree_status_t iree_hal_hip_device_queue_alloca(
+  queue_affinity_mask = queue_affinity_mask | (queue_affinity_mask - 1);
+  queue_affinity &= queue_affinity_mask;
+
+  int device_ordinal = iree_math_count_trailing_zeros_u64(queue_affinity);
+  queue_affinity = (uint64_t)1 << device_ordinal;
 
-  if (device->supports_memory_pools &&
-      !iree_all_bits_set(params.type, IREE_HAL_MEMORY_TYPE_HOST_VISIBLE)) {
+  iree_status_t status = iree_ok_status();
+  if (!iree_all_bits_set(params.type, IREE_HAL_MEMORY_TYPE_HOST_VISIBLE) &&
+      (device->supports_memory_pools ||
+       iree_hal_hip_allocator_isa(iree_hal_device_allocator(base_device)))) {
     iree_hal_buffer_t* buffer = NULL;
 
-    IREE_RETURN_IF_ERROR(iree_hal_hip_memory_pools_prepare_buffer(
-        &device->memory_pools, device->hip_dispatch_stream, pool, params,
-        allocation_size, &buffer));
+    status = iree_hal_hip_device_pepare_async_alloc(device, params,
+                                                    allocation_size, &buffer);
 
-    iree_status_t status = iree_hal_deferred_work_queue_enqueue_alloc(
-        device->work_queue, wait_semaphore_list, signal_semaphore_list, buffer);
+    iree_hal_hip_device_semaphore_buffer_operation_callback_data_t*
+        callback_data = NULL;
     if (iree_status_is_ok(status)) {
-      status = iree_hal_deferred_work_queue_issue(device->work_queue);
+      status = iree_hal_hip_device_make_buffer_callback_data(
+          device, device->host_allocator, queue_affinity, wait_semaphore_list,
+          signal_semaphore_list, buffer,
+          IREE_HAL_HIP_DEVICE_SEMAPHORE_OPERATION_ASYNC_ALLOC, &callback_data);
     }
-    if (iree_status_is_ok(status)) {
+    if (iree_status_is_ok(status) && wait_semaphore_list.count == 0) {
+      status = iree_hal_hip_dispatch_thread_add_dispatch(
+          device->topology.devices[device_ordinal].dispatch_thread,
+          &iree_hal_hip_device_perform_buffer_operation_now, callback_data);
       *out_buffer = buffer;
-    } else {
-      iree_hal_hip_buffer_set_allocation_empty(buffer);
-      iree_hal_resource_release(&buffer->resource);
+      IREE_TRACE_ZONE_END(z0);
+      return status;
     }
-    return status;
-  } else if (!iree_all_bits_set(params.type,
-                                IREE_HAL_MEMORY_TYPE_HOST_VISIBLE) &&
-             iree_hal_hip_allocator_isa(
-                 iree_hal_device_allocator(base_device))) {
-    iree_hal_buffer_t* buffer = NULL;
-
-    IREE_RETURN_IF_ERROR(iree_hal_hip_device_pepare_async_alloc(
-        device, params, allocation_size, &buffer));
 
-    iree_status_t status = iree_hal_deferred_work_queue_enqueue_alloc(
-        device->work_queue, wait_semaphore_list, signal_semaphore_list, buffer);
     if (iree_status_is_ok(status)) {
-      status = iree_hal_deferred_work_queue_issue(device->work_queue);
+      for (iree_host_size_t i = 0;
+           i < wait_semaphore_list.count && iree_status_is_ok(status); ++i) {
+        iree_hal_resource_retain(wait_semaphore_list.semaphores[i]);
```

Comment:
this makes it too difficult to tell who owns what when - only code that owns a resource should retain or release it and we need to make sure we don't double release or fail to release - I think this is broken if there's an error as it's not possible to know in cleanup which semaphores have been retained and which haven't. it's better to have a single step that goes and retains all semaphores such that after that point all semaphores can be released in one go safely **or** semaphores should only be retained when they are added to a list. having a list with only some resources half-initialized is a big source of bugs.